webpack/assets/javascripts/react_app/redux/API/APIReducer.js

```diff
@@ -0,0 +1,41 @@
+import Immutable from 'seamless-immutable';
+import { actionTypeGenerator } from './APIActionTypeGenerator';
+import { STATUS } from '../../constants';
+
+const initialState = Immutable({});
+
+const apiReducer = (state = initialState, action) => {
+  const { type, key, payload: { error, ...payload } = {} } = action;
+  if (key === undefined) return state;
+  const { REQUEST, FAILURE, SUCCESS } = actionTypeGenerator(key);
+  switch (type) {
+    case REQUEST:
+      return state.merge({
+        [key]: {
+          ...state[key],
+          ...payload,
```

Comment:
It's confusing, how the state will actually look like after `...payload`? It feels risky.

It feels to me like we need to have a proper convention so the state will be predictable.
I afraid to end up with a state that looks like:
```js
{
  status: 'pending',
  error: null,
  id: '1',
  search: 'some-search',
  page: 2,
  perPage: 20,
  items: ['item-1', 'item-2'],
}
```

For example, it doesn't clear if the `page` is the **parameter** I sent to the API or something I **received** in the API response.

I would prefer a predictable state, something like:
```js
{
  status: 'pending',
  error: null,
  params: {
    id: '1',
    search: 'some-search',
  },
  results: {
    page: 2,
    perPage: 20,
    items: ['item-1', 'item-2'],
  }
}
```


Maybe it even makes sense of having different fields for `urlParams` and `bodyParams`?

Reply:
The only problem is that there are already components that count on the payload to contain payload and data mixed.. I will open a PR to fixed that and see if we really need to keep track on the params since usually they are the props received by the component

Reply:
opened #7301 that refactor all of the mixed payloads.